### PR TITLE
[WIP] sternheimer

### PR DESCRIPTION
**Pull request by horizon04**

### CODE REVIEW

**Comment by JensWehner** [comment]
Should be merged against electrostatics

**Comment by junghans** [comment]
Changed the base.

**Comment by JensWehner** [comment]
thx

File: src/libxtp/sternheimer.cc [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+  std::complex<double> c(0, 1 / (2 * pi));
+
+  return c * SE;
+}
+
+Eigen::MatrixXcd Sternheimer::CalculateSelfEnergy(Orbitals& orbitals, double w,
+                                                  double eta,
+                                                  Eigen::VectorXd grid) {
+
+  Eigen::MatrixXcd SEC = SelfEnergyC(orbitals, grid, w, eta);
```
Comment (JensWehner) [inline src/libxtp/sternheimer.cc]: just return `SelfEnergyC(orbitals, grid, w, eta)+SelfEnergyEx(orbitals)`

File: src/libxtp/sternheimer.cc [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+//
+//        for(int n=0;n<mo_coefficients.cols();n++){
+//
+//            qpc(n)=mo_coefficients.col(n).transpose()*(SelfEnergy-VXC)*mo_coefficients.col(n);
+//
+//
+//        }
+//        return qpc;
+//
+//    }
```
Comment (JensWehner) [inline src/libxtp/sternheimer.cc]: remove commented out code

File: src/libxtp/sternheimer.cc [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+
+Eigen::MatrixXcd Sternheimer::CalculateKSselfenergie(
+    Eigen::MatrixXcd Selfenergie, Eigen::MatrixXd XC) {}
+
+Eigen::MatrixXd DipoleMatrix() {}
```
Comment (JensWehner) [inline src/libxtp/sternheimer.cc]: Delete if you do not need it

File: src/libxtp/sternheimerw.cc [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+
+#include <ctime>
+#include <eigen3/Eigen/src/Core/Block.h>
+#include <eigen3/Eigen/src/Core/Matrix.h>
+#include <eigen3/Eigen/src/Core/PlainObjectBase.h>
+#include <eigen3/Eigen/src/Eigenvalues/GeneralizedEigenSolver.h>
+#include <eigen3/Eigen/src/Eigenvalues/SelfAdjointEigenSolver.h>
+#include <eigen3/Eigen/src/IterativeLinearSolvers/BiCGSTAB.h>
+#include <eigen3/Eigen/src/IterativeLinearSolvers/ConjugateGradient.h>
+#include <eigen3/Eigen/src/IterativeLinearSolvers/IterativeSolverBase.h>
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: remove

File: src/libxtp/sternheimerw.cc [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+void SternheimerW::initializeMultishift(int basis_size) {
+
+  Multishift multi;
+  this->_multishift = multi;
+
+  _multishift.setBasisSize(basis_size);
+}
+void SternheimerW::initializePade(int size) {
+
+  PadeApprox pade;
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: as above

File: src/libxtp/sternheimerw.cc [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+Eigen::VectorXcd SternheimerW::SternheimerRHS(Eigen::MatrixXcd overlap,
+                                              Eigen::MatrixXcd density,
+                                              Eigen::MatrixXcd pertubation,
+                                              Eigen::VectorXcd coeff) {
+
+  Eigen::MatrixXcd I =
+      Eigen::MatrixXcd::Identity(overlap.rows(), overlap.cols());
+  // Calc RHS
+  Eigen::MatrixXcd t = -1 * (I - overlap * density);
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: see comment earlier about adding one to diagonal

File: src/libxtp/sternheimerw.cc [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+
+Eigen::MatrixXcd SternheimerW::SternheimerLHS(Eigen::MatrixXcd hamiltonian,
+                                              Eigen::MatrixXcd overlap,
+                                              double eps,
+                                              std::complex<double> omega,
+                                              bool pm) {
+
+  Eigen::MatrixXcd S;
+  // distinguish between +w and -w
+  if (pm == true) {
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: see earlier. This code is to a good part a copy of the sternheimer.cc don't be afraid to delete code and files as long as it is all under version control

File: src/libxtp/sternheimerw.cc [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+
+std::vector<Eigen::MatrixXcd> SternheimerW::DeltaNOSOP(
+    std::vector<std::complex<double>> w, Eigen::Vector3d r) {
+
+  Eigen::MatrixXcd V = CoulombMatrix(r);
+
+  double alpha = 2 * (_mo_energies(_num_occ_lvls) - _mo_energies(2));
+
+  std::vector<Eigen::MatrixXcd> delta_c_p;
+  std::vector<Eigen::MatrixXcd> delta_c_m;
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: better names

File: src/libxtp/sternheimerw.cc [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+
+Eigen::VectorXcd SternheimerW::SternheimerSolve(Eigen::MatrixXcd& LHS,
+                                                Eigen::VectorXcd& RHS) {
+
+  // Eigen::VectorXcd x=_multishift.CBiCG(LHS, RHS);
+
+  return _multishift.CBiCG(LHS, RHS);
+}
+
+std::vector<Eigen::MatrixXcd> SternheimerW::DeltaNOSOP(
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: better function name

File: src/libxtp/sternheimerw.cc [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+              2 * _mo_coefficients(i, v) * delta_c_p.at(m)(j, v) +
+              2 * _mo_coefficients(i, v) * delta_c_m.at(m)(j, v);
+        }
+      }
+    }
+  }
+  return delta_n;
+}
+
+std::vector<Eigen::MatrixXcd> SternheimerW::DeltaNOS(std::complex<double> w,
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: better Name

File: src/libxtp/sternheimerw.cc [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+  // Setting up container for LHS and RHS of the system
+  std::vector<Eigen::MatrixXcd> LHS_P;
+  std::vector<Eigen::MatrixXcd> LHS_M;
+  Eigen::VectorXcd RHS = Eigen::VectorXcd::Zero(basis_size);
+
+  // H_new is used if omega=0 to shift the eigenvalues away from 0
+  Eigen::MatrixXcd H_new = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+  double alpha = 2 * (mo_energies(num_occ_lvls) - mo_energies(2));
+
+  double pi = tools::conv::Pi;
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: nice

File: src/libxtp/sternheimerw.cc [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+
+    for (int v = 0; v < basis_size; v++) {
+      for (int i = 0; i < basis_size; i++) {
+        for (int j = 0; j < basis_size; j++) {
+          delta_n(i, j) = delta_n(i, j) +
+                          2 * mo_coefficients(i, v) * delta_c_p(j, v) +
+                          2 * mo_coefficients(i, v) * delta_c_m(j, v);
+        }
+      }
+    }
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: matrix product

File: src/libxtp/sternheimerw.cc [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+      H_new = H + alpha * S * p.transpose();
+      LHS_P.push_back(SternheimerLHS(H_new, S, mo_energies(v), w, true));
+      LHS_M.push_back(SternheimerLHS(H_new, S, mo_energies(v), w, false));
+    } else {
+      LHS_P.push_back(SternheimerLHS(H, S, mo_energies(v), w, true));
+      LHS_M.push_back(SternheimerLHS(H, S, mo_energies(v), w, false));
+    }
+  }
+  // Iterate over all spacial gridpoints
+  for (int r = 0; r < gridpoints.size(); r++) {
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: you can write this as a range based for loop .i.e 
`for( const std::pair<double,Eigen::Vector3d*>& gridpoint:gridpoints){`
`gridpoint.second`
`}`

File: src/libxtp/sternheimerw.cc [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+          delta_n(i, j) = delta_n(i, j) +
+                          2 * mo_coefficients(i, v) * delta_c_p(j, v) +
+                          2 * mo_coefficients(i, v) * delta_c_m(j, v);
+        }
+      }
+    }
+
+    result.push_back(delta_n);
+    delta_n = Eigen::MatrixXcd::Zero(basis_size, basis_size);
+
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: if you have to reset a matrix that often indicates that you should use it in a smaller scope, i.e. declare it inside the for loop

File: src/libxtp/sternheimerw.cc [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+  }
+
+  return result;
+}
+
+std::vector<Eigen::MatrixXcd> SternheimerW::Polarisability(
+    std::vector<std::complex<double>> grid_w,
+    std::vector<std::complex<double>> w, std::string gridtype) {
+
+  initializePade(3);
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: why 3?

File: src/libxtp/sternheimerw.cc [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+  double rp_x = 0;
+  Eigen::Vector3d gridcont = Eigen::Vector3d::Zero();
+
+  AOBasis basis = _orbitals.SetupDftBasis();
+  AODipole dipole;
+  dipole.Fill(basis);
+
+  std::cout << "Dipole integral complete" << std::endl;
+
+  for (int r = 0; r < grid.size(); r++) {
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: range based for loop?

File: src/libxtp/sternheimerw.cc [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+
+    delta_n = DeltaNOSOP(grid_w, *grid.at(r).second);
+
+    _pade.clear();
+
+    for (int o = 0; o < grid_w.size(); o++) {
+      if (r == 0) Polar.push_back(Eigen::MatrixXcd::Zero(3, 3));
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: Why? and why here?what is special about the first gridpoint?

File: src/libxtp/sternheimerw.cc [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+          for (int n = 0; n < basis_size; n++) {
+            for (int m = 0; m < basis_size; m++) {
+              gridcont = *grid.at(r).second;
+              r_x = gridcont(i);
+              Polar.at(o)(i, j) = Polar.at(o)(i, j) +
+                                  delta_n.at(o)(n, m) * r_x * grid.at(r).first *
+                                      dipole.Matrix()[j](n, m);
+            }
+          }
+        }
```
Comment (JensWehner) [inline src/libxtp/sternheimerw.cc]: there is at least one and maybe two matrix products in here ....

File: src/libxtp/sternheimerw.cc [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+                                                const Eigen::VectorXcd& RHS){
+  return LHS.colPivHouseholderQr().solve(RHS);
+}
+
+std::vector<Eigen::MatrixXcd> SternheimerW::DeltaNOneShot(
+    std::vector<std::complex<double>> w, Eigen::Vector3d r) const{
+
+  Eigen::MatrixXcd V = CoulombMatrix(r);
+  
+  double alpha = 2 * (_mo_energies(_num_occ_lvls) - _mo_energies(2));
```
Comment (junghans) [inline src/libxtp/sternheimerw.cc]: Codacy Issue found: [Variable 'alpha' is assigned a value that is never used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4286864)

File: src/libxtp/sternheimer.cc [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+//            RHS(i)=0.0;   
+//        }
+//    }
+    return RHS;
+}
+
+std::vector<Eigen::MatrixXcd> Sternheimer::DeltaNOneShot(
+    std::vector<std::complex<double>> w,const Eigen::MatrixXd& pertubation) const{
+  
+  double alpha = 2 * (_mo_energies(_num_occ_lvls) - _mo_energies(2));
```
Comment (junghans) [inline src/libxtp/sternheimer.cc]: Codacy Issue found: [Variable 'alpha' is assigned a value that is never used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4286864)

File: src/libxtp/sternheimerw.cc (hunk 19 above)
Comment (junghans) [inline src/libxtp/sternheimerw.cc]: Codacy Issue found: [Variable 'alpha' is assigned a value that is never used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4286864)

File: src/libxtp/sternheimer.cc (hunk 20 above)
Comment (junghans) [inline src/libxtp/sternheimer.cc]: Codacy Issue found: [Variable 'alpha' is assigned a value that is never used.](https://app.codacy.com/app/votca-package/xtp/pullRequest?prid=4286864)

**Comment by junghans** [comment]
Now #592

**Pull request closed without merging**